### PR TITLE
Addressing memory leaks with multi-server checkins

### PR DESCRIPTION
**Pull request by nithinj**
#### Describe Bug or Feature
Addressing the following memory leaks:

8 bytes in 1 blocks are definitely lost in loss record 8 of 71
   at 0x4C29133: malloc (vg_replace_malloc.c:299)
   by 0x5728929: strdup (in /lib64/libc-2.19.so)
   by 0x423756: parse_psi (pbs_loadconf.c:320)
   by 0x4266A5: __pbs_loadconf (pbs_loadconf.c:1103)
   by 0x427B12: pbs_loadconf (ifl_impl.c:324)
   by 0x4062DF: main (pbs_comm.c:606)

24 bytes in 1 blocks are definitely lost in loss record 1,274 of 2,146
   at 0x4C28C23: malloc (vg_replace_malloc.c:299)
   by 0x50D19E: pbs_asprintf_format (misc_utils.c:529)
   by 0x50D300: pbs_asprintf (misc_utils.c:570)
   by 0x4736F5: gen_svr_inst_id (multi_svr.c:304)
   by 0x473C4F: initialize_pbsnode (node_func.c:331)
   by 0x4781EE: shallow_vnode_dup (node_manager.c:1123)
   by 0x4784BC: set_vnode_state (node_manager.c:1212)
   by 0x477220: set_all_state (node_manager.c:547)
   by 0x47F97B: is_request (node_manager.c:4815)
   by 0x48BA22: do_tpp (pbsd_main.c:311)
   by 0x48BABE: tpp_request (pbsd_main.c:355)
   by 0x520EB0: process_socket (net_server.c:511)

48 bytes in 2 blocks are definitely lost in loss record 1,345 of 2,186
   at 0x4C28C23: malloc (vg_replace_malloc.c:299)
   by 0x50D19E: pbs_asprintf_format (misc_utils.c:529)
   by 0x50D300: pbs_asprintf (misc_utils.c:570)
   by 0x4736F5: gen_svr_inst_id (multi_svr.c:304)
   by 0x473C4F: initialize_pbsnode (node_func.c:331)
   by 0x4781EE: shallow_vnode_dup (node_manager.c:1123)
   by 0x4784BC: set_vnode_state (node_manager.c:1212)
   by 0x483AB7: set_nodes (node_manager.c:6803)
   by 0x4BA3B9: assign_hosts (req_runjob.c:1892)
   by 0x4BA1EB: where_to_runjob (req_runjob.c:1824)
   by 0x4B78BC: req_runjob (req_runjob.c:578)
   by 0x490885: dispatch_request (process_request.c:963)

50 bytes in 2 blocks are definitely lost in loss record 1,337 of 2,175
   at 0x4C29133: malloc (vg_replace_malloc.c:299)
   by 0x516CD2: pbs_asprintf_format (misc_utils.c:529)
   by 0x516E34: pbs_asprintf (misc_utils.c:570)
   by 0x47D2A1: gen_svr_inst_id (multi_svr.c:304)
   by 0x47D7FA: initialize_pbsnode (node_func.c:331)
   by 0x481D97: shallow_vnode_dup (node_manager.c:1123)
   by 0x482065: set_vnode_state (node_manager.c:1212)
   by 0x4847A9: deallocate_job_from_node (node_manager.c:2670)
   by 0x48DF4B: free_nodes (node_manager.c:7013)
   by 0x4A1A0C: rel_resc (req_jobobit.c:207)
   by 0x49F5C1: req_deletejob2 (req_delete.c:897)
   by 0x49E29C: req_deletejob (req_delete.c:449)

#### Describe Your Change
* parse_psi will free the variable first to handle the pbs conf reload case.
* shallow_vnode_dup will free attributes before assigning a value on them

#### Attach Test and Valgrind Logs/Output
[valgrind.log](https://github.com/openpbs/openpbs/files/5821837/valgrind.log)

Regression Results:
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5785|3852|2|1|0|1|3848|

Description: Rerun Only Failed Tests From #5785
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5789|3|0|0|0|0|3|

### CODE REVIEW

File: src/server/node_manager.c [hunk 1]
```diff
@@ -1089,7 +1089,13 @@ get_vnode_state_op(enum vnode_state_op op)
...
 shallow_vnode_free(struct pbsnode *vnode)
 {
+	int i;
+
 	if (vnode) {
+		free(vnode->nd_moms);
+		for (i = 0; i < ND_ATR_LAST; i++) {
+			node_attr_def[i].at_free(&vnode->nd_attr[i]);
+		}
 		free(vnode);
```
Comment (agrawalravi90) [inline src/server/node_manager.c:1099]: this seems like a pretty thorough clean, do we still want to call this function "shallow" vnode free ?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]
minor comment

File: src/server/node_manager.c (hunk 1 above)
Comment (nithinj) [reply src/server/node_manager.c]: I've reused `free_pnode` and removed `shallow_vnode_dup`. Please have a look.

**Comment by nithinj** [comment]
Requesting reviews from @suresh-thelkar

File: src/server/req_manager.c [hunk 2]
```diff
@@ -2767,13 +2767,11 @@ create_pbs_node2(char *objname, svrattrl *plist, int perms, int *bad, struct pbs
 			pbsndlist[svr_totnodes++] = pnode;
 		} else {
 			free_pnode(pnode);
-			free(pname);
```
Comment (agrawalravi90) [inline src/server/req_manager.c:2770]: why did you remove this? i don't see pname getting free'd anywhere else

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]
Please run a regression with valgrind once

File: src/server/node_manager.c [hunk 3]
```diff
@@ -1133,11 +1118,11 @@ shallow_vnode_dup(struct pbsnode *vnode)
...
 	 * Copy vnode elements (same order as "struct pbsnode" element definition)
 	 */
-	if (vnode_dup->nd_moms) {
-		/* free the initialize_pbsnode() allocation before assigning */
-		free(vnode_dup->nd_moms);
-		vnode_dup->nd_moms = NULL;
-	}
+	/* free the initialize_pbsnode() allocation before assigning */
+
+	free(vnode->nd_name);
```
Comment (agrawalravi90) [inline src/server/node_manager.c:1123]: just trying to understand, why do we need to free 'vnode' before duping it? and why are we only freeing nd_name?

File: src/server/node_manager.c [hunk 4]
```diff
@@ -1133,11 +1118,11 @@ shallow_vnode_dup(struct pbsnode *vnode)
...
+	vnode->nd_name = vnode->nd_name;
```
Comment (agrawalravi90) [inline src/server/node_manager.c:1124]: vnode->nd_name = NULL; ?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/server/node_manager.c [hunk 5]
```diff
@@ -1118,11 +1118,11 @@ shallow_vnode_dup(struct pbsnode *vnode)
...
 	 */
-	if (vnode_dup->nd_moms) {
-		/* free the initialize_pbsnode() allocation before assigning */
-		free(vnode_dup->nd_moms);
-		vnode_dup->nd_moms = NULL;
-	}
+	/* free the initialize_pbsnode() allocation before assigning */
+
+	free(vnode_dup->nd_name);
+	vnode_dup->nd_name = vnode->nd_name;
```
Comment (agrawalravi90) [inline src/server/node_manager.c:1124]: vnode_dup->nd_name is already set to strdup(vnode->nd_name) from the call to initialize_pbsnode(), why free it and set it again (and without strdup)?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/server/node_manager.c (hunk 5 above)
Comment (nithinj) [reply src/server/node_manager.c:1124]: The function does a shallow copy of the existing node. The earlier strdup is required so that we can call free_pnode on failure which would otherwise free the original copy.

**Comment by nithinj** [comment]
@agrawalravi90 ,
I've attached the valgrind logs. Please have a look.

File: src/server/node_manager.c (hunk 5 above)
Comment (agrawalravi90) [reply src/server/node_manager.c:1124]: ok .. why not just pass NULL to initialize_pbsnode() for the vnode name argument?

Comment (nithinj) [reply src/server/node_manager.c:1124]: that should work, thanks!

Comment (nithinj) [reply src/server/node_manager.c:1123]: I've addressed the comments. Please have a look.

File: src/server/node_manager.c [hunk 6]
```diff
@@ -1104,25 +1104,16 @@ shallow_vnode_dup(struct pbsnode *vnode)
...
 	 * Allocate and initialize vnode_o, then copy vnode elements into vnode_o
 	 */
-	if ((vnode_dup = malloc(sizeof(struct pbsnode)))) {
-		if (initialize_pbsnode(vnode_dup, strdup(vnode->nd_name), NTYPE_PBS) != PBSE_NONE) {
-			log_err(PBSE_INTERNAL, __func__, "vnode_dup initialization failed");
-			free_pnode(vnode_dup);
-			return NULL;
-		}
-	} else {
+	if ((vnode_dup = calloc(1, sizeof(struct pbsnode))) == NULL) {
```
Comment (agrawalravi90) [inline src/server/node_manager.c:1107]: you removed the call to initialize_pbsnode altogether, was that intentional?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/server/node_manager.c (hunk 6 above)
Comment (nithinj) [reply src/server/node_manager.c:1107]: Yes, these initialized values are getting freed and re-assigned.

**Comment by suresh-thelkar** [comment]
code changes look good to me.

**Comment by agrawalravi90** [comment]
@nithinj  please let me know once you are done with regressions, then ill take it forward

**Comment by nithinj** [comment]
@agrawalravi90 ,
I've updated the regression results in the description. It is consistent with mainline. Please have a look.

**Review by agrawalravi90 — APPROVED** [review]
Thanks for being thorough with testing
